MaxText/tests/attention_test.py

```diff
@@ -1,110 +1,402 @@
+    np.testing.assert_allclose(out, expected, atol=1e-6)
+
+  def test_multiquery_attention(self):
+    """Test MultiQueryDotProductAttention."""
+
+    # b: batch,, t: q_len, n: num_head, d: head_dim, s: kv_len
+    b = self.global_batch_size
+    t = self.max_target_length
+    n = self.num_heads
+    d = self.head_dim
+    s = self.max_target_length
+    f = self.embed_dim
+    jnp_dtype = self.dtype
+
+    np.random.seed(0)
+    query_kernel = jnp.asarray(np.random.randn(f, n, d), dtype=jnp_dtype)
+    key_kernel = jnp.asarray(np.random.randn(f, d), dtype=jnp_dtype)
+    value_kernel = jnp.asarray(np.random.randn(f, d), dtype=jnp_dtype)
+    out_kernel = jnp.asarray(np.random.randn(n, d, f), dtype=jnp_dtype)
+
+    params = {
+        'query': {
+            'kernel': query_kernel
+        },
+        'key': {
+            'kernel': key_kernel
+        },
+        'value': {
+            'kernel': value_kernel
+        },
+        'out': {
+            'kernel': out_kernel
+        }
+    }
+    inputs_q = jnp.asarray(np.random.randn(b, t, f), dtype=jnp_dtype)
+    inputs_kv = jnp.asarray(np.random.randn(b, s, f), dtype=jnp_dtype)
+
+    mqa_attention = MultiQueryDotProductAttention(
+        num_heads=self.num_heads,
+        head_dim=self.head_dim,
+        mesh=self.mesh,
+        dtype=jnp_dtype,
+        dropout_rate=self.cfg.dropout_rate,
+        name='self_attention',
     )
-    self.assertTrue(jax.numpy.allclose(flash_output, mha_output, rtol=1e-01, atol=1e-01, equal_nan=False))
 
+    _, _, _, decoder_positions = self.get_data(jnp_dtype)
+
+    out = mqa_attention.apply(
+        {'params': freeze(params)},
+        inputs_q,
+        inputs_kv,
+        decoder_segment_ids=None,
+        inputs_positions=decoder_positions,
+        mask=None,
+        bias=None,
+        deterministic=True,
+        decode=False,
+        rngs={'aqt': self.rng})
+
+    query = jnp.einsum('btf,fnd->btnd', inputs_q, query_kernel)
+    query = LLaMARotaryEmbedding(
+        embedding_dims=self.head_dim,
+        name='query_rotary',
+    )(inputs=query, position=decoder_positions)
+
+    key = jnp.einsum('bsf,fd->bsd', inputs_kv, key_kernel)
+    key = jnp.expand_dims(key, axis=-2)
+    key = LLaMARotaryEmbedding(
+        embedding_dims=self.head_dim,
+        name='key_rotary',
+    )(inputs=key, position=decoder_positions)
+    key = jnp.squeeze(key, axis=-2)
+    value = jnp.einsum('bsf,fd->bsd', inputs_kv, value_kernel)
+
+    logits = jnp.einsum('btnd,bsd->bnts', query, key)
+    weights = jax.nn.softmax(logits, axis=-1)
+    combined = jnp.einsum('bnts,bsd->btnd', weights, value)
+    expected = jnp.einsum('btnd,ndf->btf', combined, out_kernel)
+
+    out = np.asarray(out).astype(np.float16)
+    expected = np.asarray(expected).astype(np.float16)
+    np.testing.assert_allclose(out, expected, atol=1e-6)
+
+  def test_groupedquery_attention(self):
+    """Test GroupedQueryDotProductAttention."""
```

Comment:
Again, I would just test that GQA recovers MHA where the keys replicated.

Reply:
same as above